File: mSTIR/stir_demo1.m
% load STIR interface library
if ~libisloaded('mstir')
    loadlibrary('mstir')
end

opengl software

try
    % direct all diagnostic printing to Matlab Command Window
    printer = stir.printerTo('stdout');
    
    %image = stir.Image('my_uniform_image_circular.hv');
    voxel_dim = [60, 60, 31];
    voxel_size = [4.44114, 4.44114, 3.375];
    image = stir.Image();
    image.initialise(voxel_dim, voxel_size)
    image.fill(1.0)

    recon = stir.OSMAPOSLReconstruction('OSMAPOSL_test_PM_QP.par');
    
    f = stir.TruncateToCylindricalFOVImageProcessor...
        (recon.get_inter_iteration_filter());

    obj = stir.PoissonLogLikelihoodWithLinearModelForMeanAndProjData...
        (recon.get_objective_function());

%     f = recon.get_inter_iteration_filter();
%     stir.TruncateToCylindricalFOVImageProcessor(f)...
%         .set_strictly_less_than_radius(false)
    f.set_strictly_less_than_radius(false)
    f.apply(image)
    f.set_strictly_less_than_radius(true)
%     stir.TruncateToCylindricalFOVImageProcessor(f)...
%         .set_strictly_less_than_radius(true)
    
    data = image.density();
    scale = 1.0/max(max(max(data)));
    figure(1000000)
    stir.show(data, scale, 10)
    drawnow

    obj.set_sensitivity_filename('RPTsens_seg3_PM.hv')
    obj.set_input_filename('Utahscat600k_ca_seg4.hs')
    prior = obj.get_prior();
    fprintf('prior penalisation factor: %f\n', prior.get_penalisation_factor())
    prior.set_penalisation_factor(0.5)
    proj = obj.get_projector_pair();
    fprintf('tangential LORs: %d\n', proj.get_matrix().get_num_tangential_LORs())
    proj.get_matrix().set_num_tangential_LORs(2)

    tic
    obj.set_up()
    recon.set_up(image)
    recon.reconstruct(image)
    toc
 
    expectedImage = stir.Image('test_image_PM_QP_6.hv');
    diff = expectedImage.diff_from(image);
    fprintf('difference from expected image: %e\n', diff)

catch err
    % display error information
    fprintf('%s\n', err.message)
    fprintf('error id is %s\n', err.identifier)
end

File: examples/stir_demo1.py
import argparse
import numpy
import pylab
import os
import sys
sys.path.append(os.environ.get('CSTIR_SRC') + '/../pSTIR')
import time

from pStir import *

parser = argparse.ArgumentParser(description = \
'''
Mutator/accessor methods demo
''')
args = parser.parse_args()

def main():

    # direct all printing to a file
    printer = printerTo('stir_demo1.txt')

    # create OSMAPOSL reconstructor
    recon = OSMAPOSLReconstruction('OSMAPOSL_test_PM_QP2.par')

    # check/redefine some parameters
    f = CylindricFilter(recon.get_inter_iteration_filter())
    f.set_strictly_less_than_radius(True)
    obj = recon.get_objective_function()
    prior = obj.get_prior()
    print('prior penalisation factor: %f' % prior.get_penalisation_factor())
    prior.set_penalisation_factor(0.001)

    # read an initial estimate for the reconstructed image from a file
    image = PETImage('my_image0.hv')

    # plot the initial image
    data = image.as_array()
    pylab.figure(1)
    pylab.imshow(data[20,:,:])
    print('Figure 1: initial image - close window to continue')
    pylab.show()

    # set up the reconstructor
    print('setting up, please wait...')
    recon.set_up(image)

    # run reconstruction
    print('reconstructing, please wait...')
    start_time = time.time()
    recon.reconstruct(image)
    elapsed_time = time.time() - start_time
    print('elapsed time: %f' % elapsed_time)

    # compare the reconstructed image to the expected image
    expectedImage = PETImage('expected_image.hv')
    diff = expectedImage.diff_from(image)
    print('difference from expected image: %e' % diff)

    # compare the reconstructed image to the exact image
    exactImage = PETImage('my_image.hv')

    # let the user inspect any z-crossections of the image they want to
    data = image.as_array()
    x_data = exactImage.as_array()
    nz = data.shape[0]
    print('Enter z-coordinate of the slice to view it')
    print('(a value outside the range [0 : %d] will stop the loop)'%(nz - 1))
    while True:
        s = str(input('z-coordinate: '))
        if len(s) < 1:
            break
        z = int(s)
        if z < 0 or z >= nz:
            break
        pylab.figure(z)
        pylab.imshow(data[z,:,:])
        pylab.figure(1000 + z)
        pylab.imshow(x_data[z,:,:])
        print('close Figure %d window, then' % (1000 + z))
        print('close Figure %d window to continue' % z)
        pylab.show()

# if anything goes wrong, an exception will be thrown 
# (cf. Error Handling section in the spec)
try:
    main()
except error as err:
    # display error information
    print('STIR exception occured: %s' % err.value)
